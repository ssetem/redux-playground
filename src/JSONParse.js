import chalk from "chalk"

export const TYPES = {
    WHITESPACE: "WHITESPACE",
    UNRECOGNISED: "UNRECOGNISED",
    NUMBER: "NUMBER",
    STRING: "STRING",
    STRING_KEY: "STRING_KEY",
    NULL: "NULL",
    BOOLEAN: "BOOLEAN",
    BRACE_START: "BRACE_START",
    BRACE_END: "BRACE_END",
    BRACKET_START: "BRACKET_START",
    BRACKET_END: "BRACKET_END",
    COMMA: "COMMA",
    COLON: "COLON",
    END: "END",
}

export const BRACKET_COLOR = chalk.white
export const COLORS = {
    [TYPES.STRING_KEY]:chalk.hsl(210, 44,47),
    [TYPES.STRING]: chalk.hsl(29,54,61),
    [TYPES.NUMBER]: chalk.hsl(95, 38, 62),
    [TYPES.NULL]:chalk.blue,
    [TYPES.BOOLEAN]:chalk.hsl(207, 82, 66),
    [TYPES.BRACE_START]:BRACKET_COLOR,
    [TYPES.BRACE_END]:BRACKET_COLOR,
    [TYPES.BRACKET_START]:BRACKET_COLOR,
    [TYPES.BRACKET_END]:BRACKET_COLOR,
    ERROR:chalk.underline.red.bold.bgHex("#333")  
}

export const TOKEN_TYPES = [
    { regex: /^\s+/, type: TYPES.WHITESPACE },
    { regex: /^-?\d+(?:\.\d+)?(?:e[+\-]?\d+)?/i, type: TYPES.NUMBER },
    { regex: /^"((?:\\.|[^"])*)"(?=\s*:)/, type: TYPES.STRING_KEY },
    { regex: /^"((?:\\.|[^"])*)"/, type: TYPES.STRING },
    { regex: /^(?:true|false)/, type: TYPES.BOOLEAN },
    { regex: /^null/, type: TYPES.NULL },
    { regex: /^\[/, type: TYPES.BRACKET_START },
    { regex: /^\]/, type: TYPES.BRACKET_END },
    { regex: /^\{/, type: TYPES.BRACE_START },
    { regex: /^\}/, type: TYPES.BRACE_END },
    { regex: /^,/, type: TYPES.COMMA },
    { regex: /^:/, type: TYPES.COLON },
    { regex: /^\w+/, type: TYPES.UNRECOGNISED },
    { regex: /^.+/, type: TYPES.UNRECOGNISED }
]

export class SourceCursor {
    constructor(currentString) {
        this.currentString = currentString
        this.cursor = 0
    }

    matchNextToken() {
        for (let tokenType of TOKEN_TYPES) {
            if (tokenType.regex.test(this.currentString)) {
                let matches = this.currentString.match(tokenType.regex)
                let value = matches[0]

                let token = {
                    type: tokenType.type,
                    value: matches[0],
                    groups: matches.slice(1),
                    cursor: this.cursor
                }

                this.cursor += value.length
                this.currentString = this.currentString.slice(value.length)
                return token
            }
        }
    }

    tokenize() {
        let tokens = []
        do {
            let token = this.matchNextToken()
            if (!token) {
                throw new Error("Could not match a token for " + this.currentString.slice(10))
            }
            tokens.push(token)
        } while (this.currentString.length > 0)
        return tokens
    }
}

export class TokensCursor {
    constructor(tokens){
        this.originalTokens = tokens
        this.tokens = tokens
    }

    take(){
        let token = this.tokens[0]
        this.tokens = this.tokens.slice(1)
        return token
    }

    hasNext(){
        return this.tokens.length > 0
    }

    takeStrictByType(type){
        let token = this.take()
        if(token.type != type){
            let errorMessage = `Expected ${type} but got ${token.type}`
            this.error(token, errorMessage)
            throw new Error(errorMessage)
        }
        return token
    }

    removeTokensByType(type){
        this.tokens = this.tokens.filter((token)=> {
            return token.type != type
        })
    }

    peek(){
        if(this.tokens.length === 0){
            return {
                type:TYPES.END                
            }
        } else {
            return this.tokens[0]
        }
    }

    highlightErrors(){

        return this.highlight(this.erroredTokens)
    }
    highlight(tokens=this.tokens){        
        return tokens.map((token)=> {
            let highlight = COLORS[token.type]
            if(token.error){
                highlight = COLORS.ERROR
            }
            return highlight ? highlight(token.value) : token.value
        }).join("")
    }

    error(erroredToken, errorMessage){
        if(erroredToken.type === TYPES.END){
            this.erroredTokens = this.originalTokens.concat({
                ...erroredToken,
                value:" ",
                error:errorMessage
            })
        } else {
            this.erroredTokens = this.originalTokens.map((token) => {
                if (erroredToken === token) {
                    return {
                        ...token,
                        error: errorMessage
                    }
                }
                return token
            })
        }
        
    }
}

export const JSONParse = (str)=>{
    let tokens = new SourceCursor(str).tokenize()
    let tokensCursor = new TokensCursor(tokens)
    tokensCursor.removeTokensByType(TYPES.WHITESPACE)
    try {
        let value = matchValue(tokensCursor)        
        if(tokensCursor.hasNext()){
            tokensCursor.error(tokensCursor.peek(), "Additional Tokens found")
            throw new Error("Additional tokens found")
        }
        return value
    } catch(e){
        console.log(e.stack)
        console.log(tokensCursor.highlightErrors(), "\n"+chalk.red(e.message))
        // console.error(e)
    }
}

export const matchValue = (tokensCursor)=> {
    let token = tokensCursor.peek()
    switch(token.type){
        case TYPES.BRACE_START:
            return matchObject(tokensCursor)
        case TYPES.BRACKET_START:
            return matchArray(tokensCursor)
        case TYPES.STRING:
            return tokensCursor.take().groups[0]
        case TYPES.NUMBER:
            return Number(tokensCursor.take().value)
        case TYPES.NULL:
            tokensCursor.take()
            return null
        case TYPES.BOOLEAN:
            return tokensCursor.take().value === 'true'
        default:
            tokensCursor.error(token, "Could not match value")
            throw new Error("Could not match value")
    }
}

export const matchObject = (tokensCursor)=> {
    tokensCursor.takeStrictByType(TYPES.BRACE_START)
    let token = tokensCursor.peek()
    let ob = {}
    if (token.type === TYPES.BRACE_END){
        tokensCursor.take()
        return ob
    }

    do {
        let key = tokensCursor.takeStrictByType(TYPES.STRING_KEY)
        tokensCursor.takeStrictByType(TYPES.COLON)
        let value = matchValue(tokensCursor)
        ob[key.groups[0]] = value
        token = tokensCursor.peek()
        if (token.type === TYPES.COMMA) {
            tokensCursor.take()
        } else if (token.type === TYPES.BRACE_END) {
            tokensCursor.take()
            return ob
        } else {
            let errorMessage = `Expected , or }`
            tokensCursor.error(token, errorMessage)
            throw new Error(errorMessage)
        }
    } while(tokensCursor.peek())
}

export const matchArray = (tokensCursor)=> {
    tokensCursor.takeStrictByType(TYPES.BRACKET_START)
    let token = tokensCursor.peek()
    let arr = []
    if(token.type === TYPES.BRACKET_END){
        tokensCursor.take()
        return arr
    }

    do {
        arr.push(matchValue(tokensCursor))
        token = tokensCursor.peek()
        if(token.type === TYPES.COMMA){
            tokensCursor.take()
        } else if(token.type === TYPES.BRACKET_END){
            tokensCursor.take()
            return arr
        } else {
            let errorMessage = `Expected , or ]`
            tokensCursor.error(token, errorMessage)
            throw new Error(errorMessage)
        }
    } while(tokensCursor.peek())
}