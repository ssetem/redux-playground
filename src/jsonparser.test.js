import { SourceCursor, TokensCursor, JSONParse } from "./JSONParse"
import fs from "fs"
import path from "path"
describe("json parser", function () {

    describe("tokenizer", () => {

        it("tokenizes numbers correctly", () => {
            let str = " 12 -12 -12.1222E12 +12.303"
            let tokens = new SourceCursor(str).tokenize()
            expect(tokens).toMatchSnapshot()
        })

        it("tokenizes strings correctly", () => {
            let str = `"\\"" "hello" "hello world"`
            let tokens = new SourceCursor(str).tokenize()
            expect(tokens).toMatchSnapshot()
        })

        it("matches null false true correctly", () => {
            let str = `null false true`
            let tokens = new SourceCursor(str).tokenize()
            expect(tokens).toMatchSnapshot()
        })

        it("matches array syntax", () => {
            let str = `[ 1, "2", 3]`
            let tokens = new SourceCursor(str).tokenize()
            expect(tokens).toMatchSnapshot()
        })

        it("matches object syntax", () => {
            let str = `{"a":1, "b":"c"}`
            let tokens = new SourceCursor(str).tokenize()
            expect(tokens).toMatchSnapshot()
        })
    })

    describe("TokensCursor", () => {
        it("peek take", () => {
            let tokens = new SourceCursor("1false2true").tokenize()
            let tokensCursor = new TokensCursor(tokens)
            expect(tokensCursor.peek().value).toEqual('1')
            expect(tokensCursor.tokens.length).toEqual(4)
            expect(tokensCursor.take().value).toEqual('1')
            expect(tokensCursor.tokens).toMatchSnapshot()
        })

        fit("test highlight", ()=> {
            let str = `{
  "name": "searchkit-starter-react16",
  "version": 1.1,
  "private": true,
  "devDependencies": {
    "enzyme": "^3.1.0",
    "enzyme-adapter-react-16": "^1.0.2",
    "js-beautify": "^1.7.4",
    "raf": "^3.4.0",
    "react-scripts": "^1.0.14"
  },
  "dependencies": {
    "react": "^16.0.0",
    "react-autocomplete": "^1.7.2",
    "react-autosuggest": "^9.3.2",
    "react-dom": "^16.0.0",
    "react-select": "^1.0.0-rc.10",
    "searchkit": "^2.3.0-7"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom",
    "eject": "react-scripts eject"
  }
}
`
            let tokens = new SourceCursor(str).tokenize()
            let tokensCursor = new TokensCursor(tokens)
            console.log(tokensCursor.highlight())
        })
    })

    describe("JSONParse", function(){

        it("parses strings", ()=> {
            expect(JSONParse(JSON.stringify("hi"))).toEqual("hi")
        })

        it("parses numbers", () => {
            expect(JSONParse(JSON.stringify(-12.12e12))).toEqual(-12.12e12)
        })

        it("parses null", ()=> {
            expect(JSONParse("null")).toEqual(null)
        })

        it("parses boolean", () => {
            expect(JSONParse("false")).toEqual(false)
        })

        it("handles whitespace", () => {
            expect(JSONParse(" false")).toEqual(false)
        })

        it("parses arrays", () => {
            expect(JSONParse("[]")).toEqual([])
            expect(JSONParse("[1, false, null]")).toEqual([1,false,null])
            // JSONParse("[1,2}")
        })

        it("parses objects", () => {
            expect(JSONParse("{}")).toEqual({})
            let ob = { "a": 1, "b": true, "c": { "d": null } }
            expect(JSONParse(JSON.stringify(ob))).toEqual(ob)
            console.log(JSONParse(`{"a":1}}`))
        })

        fit("parses big object", ()=> {
            let bigJsonStr = fs.readFileSync(path.join(__dirname + "/bigJson.txt")).toString()
            console.log(JSONParse(bigJsonStr))
        })

        // fit("parses additional tokens", ()=> {
        //     console.log(JSONParse(`{"a":2`))
        // })
    })
})