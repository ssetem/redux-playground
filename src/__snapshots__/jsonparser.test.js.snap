// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json parser TokensCursor peek take 1`] = `
Array [
  Object {
    "cursor": 1,
    "groups": Array [],
    "type": "BOOLEAN",
    "value": "false",
  },
  Object {
    "cursor": 6,
    "groups": Array [],
    "type": "NUMBER",
    "value": "2",
  },
  Object {
    "cursor": 7,
    "groups": Array [],
    "type": "BOOLEAN",
    "value": "true",
  },
]
`;

exports[`json parser tokenizer matches array syntax 1`] = `
Array [
  Object {
    "cursor": 0,
    "groups": Array [],
    "type": "BRACKET_START",
    "value": "[",
  },
  Object {
    "cursor": 1,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 2,
    "groups": Array [],
    "type": "NUMBER",
    "value": "1",
  },
  Object {
    "cursor": 3,
    "groups": Array [],
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "cursor": 4,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 5,
    "groups": Array [
      "2",
    ],
    "type": "STRING",
    "value": "\\"2\\"",
  },
  Object {
    "cursor": 8,
    "groups": Array [],
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "cursor": 9,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 10,
    "groups": Array [],
    "type": "NUMBER",
    "value": "3",
  },
  Object {
    "cursor": 11,
    "groups": Array [],
    "type": "BRACKET_END",
    "value": "]",
  },
]
`;

exports[`json parser tokenizer matches null false true correctly 1`] = `
Array [
  Object {
    "cursor": 0,
    "groups": Array [],
    "type": "NULL",
    "value": "null",
  },
  Object {
    "cursor": 4,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 5,
    "groups": Array [],
    "type": "BOOLEAN",
    "value": "false",
  },
  Object {
    "cursor": 10,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 11,
    "groups": Array [],
    "type": "BOOLEAN",
    "value": "true",
  },
]
`;

exports[`json parser tokenizer matches object syntax 1`] = `
Array [
  Object {
    "cursor": 0,
    "groups": Array [],
    "type": "BRACE_START",
    "value": "{",
  },
  Object {
    "cursor": 1,
    "groups": Array [
      "a",
    ],
    "type": "STRING",
    "value": "\\"a\\"",
  },
  Object {
    "cursor": 4,
    "groups": Array [],
    "type": "COLON",
    "value": ":",
  },
  Object {
    "cursor": 5,
    "groups": Array [],
    "type": "NUMBER",
    "value": "1",
  },
  Object {
    "cursor": 6,
    "groups": Array [],
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "cursor": 7,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 8,
    "groups": Array [
      "b",
    ],
    "type": "STRING",
    "value": "\\"b\\"",
  },
  Object {
    "cursor": 11,
    "groups": Array [],
    "type": "COLON",
    "value": ":",
  },
  Object {
    "cursor": 12,
    "groups": Array [
      "c",
    ],
    "type": "STRING",
    "value": "\\"c\\"",
  },
  Object {
    "cursor": 15,
    "groups": Array [],
    "type": "BRACE_END",
    "value": "}",
  },
]
`;

exports[`json parser tokenizer tokenizes numbers correctly 1`] = `
Array [
  Object {
    "cursor": 0,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 1,
    "groups": Array [],
    "type": "NUMBER",
    "value": "12",
  },
  Object {
    "cursor": 3,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 4,
    "groups": Array [],
    "type": "NUMBER",
    "value": "-12",
  },
  Object {
    "cursor": 7,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 8,
    "groups": Array [],
    "type": "NUMBER",
    "value": "-12.1222E12",
  },
  Object {
    "cursor": 19,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 20,
    "groups": Array [],
    "type": "NUMBER",
    "value": "+12.303",
  },
]
`;

exports[`json parser tokenizer tokenizes strings correctly 1`] = `
Array [
  Object {
    "cursor": 0,
    "groups": Array [
      "\\\\\\"",
    ],
    "type": "STRING",
    "value": "\\"\\\\\\"\\"",
  },
  Object {
    "cursor": 4,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 5,
    "groups": Array [
      "hello",
    ],
    "type": "STRING",
    "value": "\\"hello\\"",
  },
  Object {
    "cursor": 12,
    "groups": Array [],
    "type": "WHITESPACE",
    "value": " ",
  },
  Object {
    "cursor": 13,
    "groups": Array [
      "hello world",
    ],
    "type": "STRING",
    "value": "\\"hello world\\"",
  },
]
`;
